requestmanager/reconciledloader/load.go

```diff
@@ -0,0 +1,178 @@
+	if !rl.mostRecentOfflineLoad.empty() {
+		// since we aren't retrying the most recent offline load, it's time to record it in the traversal record
+		rl.traversalRecord.RecordNextStep(
+			rl.mostRecentOfflineLoad.linkContext.LinkPath.Segments(),
+			rl.mostRecentOfflineLoad.link.(cidlink.Link).Cid,
+			rl.mostRecentOfflineLoad.successful,
+		)
+		rl.mostRecentOfflineLoad = loadAttempt{}
+	}
+
+	// the private method does the actual loading, while this wrapper simply does the record keeping
+	remoteOnline, result := rl.blockReadOpener(lctx, link)
+
+	// now, either record cause we're online or cache to allow a retry if we're offline
+	if remoteOnline {
+		rl.traversalRecord.RecordNextStep(lctx.LinkPath.Segments(), link.(cidlink.Link).Cid, result.Err == nil)
+	} else {
+		rl.mostRecentOfflineLoad.link = link
+		rl.mostRecentOfflineLoad.linkContext = lctx
+		rl.mostRecentOfflineLoad.successful = result.Err == nil
+	}
+	return result
+}
+
+func (rl *ReconciledLoader) blockReadOpener(lctx linking.LinkContext, link datamodel.Link) (remoteOnline bool, result types.AsyncLoadResult) {
+
+	// catch up the remore or determine that we are offline
+	hasRemoteData, err := rl.waitRemote()
+	if err != nil {
+		return hasRemoteData, types.AsyncLoadResult{Err: err, Local: !hasRemoteData}
+	}
+
+	// if we're offline just load local
+	if !hasRemoteData {
+		return false, rl.loadLocal(lctx, link)
+	}
+
+	// only attempt remote load if after reconciliation we're not on a missing path
+	if !rl.pathTracker.stillOnMissingRemotePath(lctx.LinkPath) {
+		data, err := rl.loadRemote(lctx, link)
+		if data != nil {
+			return true, types.AsyncLoadResult{Data: data, Local: false}
+		}
+		if err != nil {
+			return true, types.AsyncLoadResult{Err: err, Local: false}
+		}
+	}
+	// remote had missing or duplicate block, attempt load local
+	return true, rl.loadLocal(lctx, link)
+}
+
+func (rl *ReconciledLoader) loadLocal(lctx linking.LinkContext, link datamodel.Link) types.AsyncLoadResult {
+	stream, err := rl.lsys.StorageReadOpener(lctx, link)
+	if err != nil {
+		return types.AsyncLoadResult{Err: graphsync.MissingBlockErr{Link: link}, Local: true}
+	}
+	// skip a stream copy if it's not needed
+	if br, ok := stream.(byteReader); ok {
+		return types.AsyncLoadResult{Data: br.Bytes(), Local: true}
+	}
+	localData, err := ioutil.ReadAll(stream)
+	if err != nil {
+		return types.AsyncLoadResult{Err: graphsync.MissingBlockErr{Link: link}, Local: true}
+	}
+	return types.AsyncLoadResult{Data: localData, Local: true}
+}
+
+func (rl *ReconciledLoader) loadRemote(lctx linking.LinkContext, link datamodel.Link) ([]byte, error) {
+	rl.lock.Lock()
+	head := rl.remoteQueue.first()
+	buffered := rl.remoteQueue.consume()
+	rl.lock.Unlock()
+
+	// verify it matches the expected next load
+	if !head.link.Equals(link.(cidlink.Link).Cid) {
+		return nil, graphsync.RemoteIncorrectResponseError{
```

Comment:
we have zero tolerance for out-of-order blocks right? so this error is final and breaks the entire transfer?

Reply:
yes and yes. Remotes had best not fuck up.

Reply:
I should probably remove swearing in a Github comment and yet I am not feeling compelled to :P